### PR TITLE
switch-to-configuration: use Net::DBus to retrieve the list of units

### PR DESCRIPTION
**Pull request by edef1c**
###### Motivation for this change

This resolves the FIXME, and opens up the possibility of using more of
the systemd DBus interface to make things more robust.

###### Things done

- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by volth** [comment]
It might break ```nixinfect```-like tools which run ```$nixclosure/bin/switch-to-configuration boot``` on Ubuntu/CentOS/.... and then reboot to NixOS. They do not need to get list of units, but it may fail at ```use Net::DBus;``` because of absence of the library. I have not tested, just thoughts.

**Comment by edef1c** [comment]
@volth The library won't be absent - it's part of the Nix store, and included by its store path, just like `File::Slurp` is.

**Comment by volth** [comment]
I meant that this script might be run using Ubuntu's perl and libraries

**Comment by edef1c** [comment]
nixos-infect just runs it from the store, using the same shebang line it would use on NixOS, and hence the same interpreter and libraries: https://github.com/elitak/nixos-infect/blob/769dce60c939add8fae5bc9f03fc96d8661a76a7/nixos-infect#L235

**Comment by copumpkin** [comment]
Looks fine to me. @edolstra might have opinions?

File: nixos/modules/system/activation/top-level.nix [hunk 1]
```diff
@@ -127,7 +127,7 @@ let
       configurationName = config.boot.loader.grub.configurationName;
 
       # Needed by switch-to-configuration.
-      perl = "${pkgs.perl}/bin/perl -I${pkgs.perlPackages.FileSlurp}/lib/perl5/site_perl";
+      perl = "${pkgs.perl}/bin/perl -I${pkgs.perlPackages.FileSlurp}/lib/perl5/site_perl -I${pkgs.perlPackages.NetDBus}/lib/perl5/site_perl -I${pkgs.perlPackages.XMLParser}/lib/perl5/site_perl -I${pkgs.perlPackages.XMLTwig}/lib/perl5/site_perl";
```
Comment (jtojnar) [inline nixos/modules/system/activation/top-level.nix]: Can be cleaned up to `${concatStringsSep " " (map (lib: "-I${lib}/${pkgs.perl.libPrefix}") (with pkgs.perlPackages; [ FileSlurp NetDBus XMLParser XMLTwig ]))}`. Is not there an existing function (like `python.withPackages`) for that?

Comment (edef1c) [reply nixos/modules/system/activation/top-level.nix]: I don't know of one, and digging around naively didn't find me it. I could most definitely write one, though.

**Comment by edef1c** [comment]
@grahamcofborg test switchTest

**Comment by matthewbauer** [comment]
Honestly looks good to me. It's usually good to merge these things before they go stale. Any objections?

@GrahamcOfBorg eval

**Comment by edef1c** [comment]
Any remaining objections?

**Comment by matthewbauer** [comment]
This is breaking something:

https://hydra.nixos.org/build/73531916/nixlog/1

Reverted in ca30c5e061e

**Pull request merged**